### DISCUSSION TITLE
基于x-validator做异步校验，form的validating状态不会自动跟随改变，校验结束时onFormValidateEnd事件也没有触发

### DISCUSSION THREAD

**Discussion by feyy** [alibaba/formily#2513] [Ideas]
#2512 
应用场景：基于x-validator属性设置的异步校验会自动触发，我只想在onFormValidateEnd时，获取到校验通过后的表单项值。所以希望form.validating状态应该能够自动识别x-validator是否处于validating中的，同时onFormValidateEnd事件在validating由true变为false时候触发。
正如issue中提到的，现在x-validator的自动校验，form的validating状态，和onFormValidateEnd事件三者是分开的。

**Comment by janryWang** [comment]
onFormValidateEnd的时机不是所有字段独立校验完成的时机，因为字段如果自己单独输入，完全是另一套生命周期，跟Form一点关系没有，所以onFormValidateEnd只是form.submit/form.validate单独触发执行的时机

**Reply by feyy** [reply]
为啥不能是关联的呢？一个form的状态不应该就是所有form下field状态的集合么？ 不应该像form的valid，invalid状态一样，field invalid的时候，form就invalid
而且你去调用form.validate()不也会把所有field的校验都跑一遍么？说不通啊？

**Reply by janryWang** [reply]
那你觉得应该怎么改合适呢

**Reply by feyy** [reply]
form的validating也应该是 所有filed的validating的集合塞，同时validating变化的时候也会相应的触发onFormValidateStart和onFormValidateEnd事件；

这样使用x-validator做远程校验时：
field.validating (true) -> form.validating(true) -> trigger form.onFormValidateStart -> 远程校验结束 -> field.validating (false) -> form.validating(false) ->  trigger form.onFormValidateEnd

整个状态流转自动进行，不需要开发者去手动修改form的validating状态

**Reply by janryWang** [reply]
那这样就会存在一个问题，Submit组件的loading状态，一闪一闪的，因为validating一下子为true，一下子为false

**Reply by feyy** [reply]
不就应该这样么？如果你要做sumit loading状态关联的话？
主要是field处于validating中的时候，你的整个form也不应该可以提交塞

**Reply by janryWang** [reply]
你硬是要把validating连接起来又不是不可以，写个effects函数就行了
```tsx
onFieldReact('*',(field,form)=>{
  if(isField(field)){
      form.validating = field.validating
  }
})
```

**Reply by janryWang** [reply]
> 不就应该这样么？如果你要做sumit loading状态关联的话？ 主要是field处于validating中的时候，你的整个form也不应该可以提交塞

提交本来就是会阻塞的呀，因为是整体校验的，我说的是输入的过程中，Submit按钮一闪一闪的，这种视觉干扰影响太大了

**Reply by feyy** [reply]
不是硬是，我觉得你们也可以内部讨论下这个问题？我觉得这才是合理的塞？也没有坏处
要不然form.validating需要手动控制，理解成本太高了。还有本身和form.validating = true，已经form.setValidating(true)的迷之关系

**Reply by feyy** [reply]
> 

我做的实时保存，没有显示的提交过程；直接响应的form的value change事件 和 validateEnd事件，里面需要根据validating和valid状态决定是否使用当前值

**Comment by janryWang** [comment]
> 还有本身和form.validating = true，已经form.setValidating(true)的迷之关系

具体说说是什么蜜汁关系

**Reply by feyy** [reply]
就是直接写form.validating不会触发onFieldValidateStart和onFieldValidateEnd事件；
通过form.setValidating()设置时，true设置无效，但能触发onFieldValidateStart和onFieldValidateEnd事件；

**Reply by janryWang** [reply]
setValidating是有截流逻辑的，validating不是立即变成true

**Comment by janryWang** [comment]
> 我做的实时保存，没有显示的提交过程；直接响应的form的value change事件 和 validateEnd事件，里面需要根据validating和valid状态决定是否使用当前值

实时保存就应该只消费onFieldValidateStart和onFieldValidateEnd

**Reply by feyy** [reply]
这不是现在onFieldValidateEnd并不会触发么

**Comment by feyy** [comment]
期望切实考虑下做实时保存的场景，怎么去处理form的值吧。
目前不去手动调用setValidating根本不会触发onFieldValidateStart和onFieldValidateEnd；然而onFormValuesChange又没法获取到真确的validating状态

**Reply by janryWang** [reply]
搞不懂，做实时提交，为啥会强依赖validating，validating一般是给UI消费的，要做提交拦截，直接在onFieldValidateSuccess里消费form.valid，然后提交不就行了么？
```

onFieldValidateSuccess('*',()=>{
  if(form.valid){
    form.submit()
  }
})

```

**Reply by feyy** [reply]
刚试了下onFieldValidateSuccess确实满足场景需要，但是onFormValidateSuccess确也不是想象中的运行方式，我先用onFieldValidateSuccess实现吧。

之所以依赖validating状态，就是配合onFieldValidateEnd，onFormValuesChange和valid状态实现onFieldValidateSuccess的效果😭
